Include erb files for analysis, following Flay

ABaldwinHunter opened the pull request:
This erb-processing code is lifted directly from flay:
https://github.com/seattlerb/flay/blob/master/lib/flay_erb.rb#L13


@codeclimate/review @noahd1

wfleming commented on lib/cc/engine/analyzers/ruby/main.rb:
```diff
@@ -33,8 +34,45 @@ def overage(issue)
+            end
+          end
+
+          class Erubis < ::Erubis::Eruby # :nodoc:
```
:lipstick: The `nodoc` comment is extraneous since don't use rdoc.

ABaldwinHunter replied:
:+1:  thanks

wfleming commented on lib/cc/engine/analyzers/ruby/main.rb:
```diff
@@ -33,8 +34,47 @@ def overage(issue)
 
           def process_file(file)
-            RubyParser.new.process(File.binread(file), file, TIMEOUT)
+            if File.extname(file) == ".erb"
```
WDYT think about extracting all of this into a `Parser` class like the other languages have? Keeping that logic here made sense when it was one-liner, but I think this is complex enough that extracting makes sense.

ABaldwinHunter replied:
I actually prefer the single file because I think the `main` class is fairly small (< 5 methods), and adding encapsulating parser files might be adding needless encapsulation.

Let me know if you feel strongly though - and I can make a change.

ABaldwinHunter commented:
For the CC duplication issue, I considered  extracting part of the method, but I think it adds more lines and complexity in this case. More readable as is.

ABaldwinHunter commented:
ping @wfleming @codeclimate/review Ready for re-review!

wfleming commented on lib/cc/engine/analyzers/ruby/main.rb:
```diff
@@ -33,7 +34,45 @@ def overage(issue)
+          end
+
+          # taken from Flay, https://github.com/seattlerb/flay/blob/master/lib/flay_erb.rb
+
```
:lipstick: extra space

ABaldwinHunter replied:
:+1:

wfleming commented on lib/cc/engine/analyzers/ruby/main.rb:
```diff
@@ -33,7 +34,45 @@ def overage(issue)
+            begin
+              RubyParser.new.process(ruby, file)
+            rescue => e
+              $stderr.puts ruby
```
This seems a bit excessive, and could be very large.

We catch exceptions up the stack, and we're raising right below here anyway: we can probably lose the begin/rescue here. We'll get the filename up the stack & us/customers can user that for debugging.

ABaldwinHunter replied:
:+1:  sounds good

wfleming commented:
Noticed 2 more small things. If we're going to exclude the duplication issue, maybe we should add it to the ignored fingerprints in config?

Otherwise, LGTM.

ABaldwinHunter commented:
@codeclimate/review 

Ready for re-review!

@wfleming I 

- removed the begin/rescue clause and output to stderr [here](https://github.com/codeclimate/codeclimate-duplication/pull/81/files#diff-415d8ecf3775cdd67402dc77177f7b15R44), as suggested
- added codeclimate issue to finger_print excludes

wfleming commented on lib/cc/engine/analyzers/ruby/main.rb:
```diff
@@ -33,7 +34,37 @@ def overage(mass)
+          end
+
+          def process_erb(file)
+            erb = File.read(file)
```
This is a minor pondering that just occurred to me: we use `binread` above but just use `read` here. I suspect this is down to file encodings. You comfortable having these be different?

ABaldwinHunter replied:
Thanks for the catch, `binread` sounds good to me.

I pulled the code from `flay` and tested it out locally without an issue, and didn't realize the intention behind using `File.binread`. Making switch

wfleming commented:
One small question, but LGTM.
